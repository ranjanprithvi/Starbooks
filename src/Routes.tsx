import React from "react";
import { Navigate, Route, Routes as ReactRoutes } from "react-router-dom";
import Books from "./components/Books/Books";
import BookDetails from "./components/Books/BookDetails";
import BookForm from "./components/Books/BookForm";
import NotFound from "./components/common/NotFound";
import Rentals from "./components/Rentals/Rentals";
import Users from "./components/Users/Users";
import RentalForm from "./components/Rentals/RentalForm";
import UserDetails from "./components/Users/UserDetails";
import UserForm from "./components/Users/UserForm";
import LoginForm from "./components/Users/LoginForm";
import ProtectedAdminComponent from "./components/common/ProtectedAdminComponent";
import ProtectedComponent from "./components/common/ProtectedComponent";
import AuthorGenreList from "./components/Authors/AuthorGenreList";
import Home from "./components/Home";

const Routes = ({ dataView }: { dataView?: string | undefined }) => {
    return (
        <ReactRoutes>
            <Route path="/" element={<Home />} />
            <Route path="/login" element={<LoginForm />} />
            <Route path="/books" element={<Books />} />
            <Route path="bookDetails/:id" element={<BookDetails />} />
            <Route
                path="books/:id"
                element={
                    <ProtectedAdminComponent>
                        <BookForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="users"
                element={
                    <ProtectedAdminComponent>
                        <Users dataView={dataView} />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="userDetails/:id"
                element={
                    <ProtectedComponent>
                        <UserDetails />
                    </ProtectedComponent>
                }
            />
            <Route
                path="users/:id"
                element={
                    <ProtectedAdminComponent>
                        <UserForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals"
                element={
                    <ProtectedAdminComponent>
                        <Rentals />
                    </ProtectedAdminComponent>
                }
            />
            <Route
                path="rentals/:id"
                element={
                    <ProtectedAdminComponent>
                        <RentalForm />
                    </ProtectedAdminComponent>
                }
            />
            <Route path="/authorGenreList" element={<AuthorGenreList />} />
            <Route path="/not-found" element={<NotFound />} />
            <Route path="*" element={<Navigate to="/not-found" replace />} />
        </ReactRoutes>
    );
};

export default Routes;
