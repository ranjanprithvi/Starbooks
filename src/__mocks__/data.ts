export const bookObj = {
    _id: "123",
    title: "test",
    author: { _id: "1", name: "author1" },
    genre: { _id: "1", name: "genre1" },
    yearPublished: 2021,
    rating: 5,
    coverImage:
        "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1681496408i/58536026.jpg",
    description: "test",
    numberInStock: 1,
};
