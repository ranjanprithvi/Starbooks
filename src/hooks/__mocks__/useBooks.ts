const useBooks = jest.fn().mockReturnValueOnce({
    books: [
        {
            _id: "65987aa9e6b72d92ddcc1754",
            title: "asdf",
            genre: {
                _id: "6363a31891156291b10c0fea",
                __v: 0,
                name: "Fantasy",
            },
            author: {
                _id: "635c226f73691fb5ae1d3eea",
                __v: 0,
                name: "Autocar",
            },
            numberInStock: 0,
            rating: 0,
            yearPublished: 1232,
            coverImage: "",
            description: "akdd",
            __v: 0,
        },
        {
            _id: "651becad552934bbabb7bdd5",
            title: "The Body in the Library",
            genre: {
                _id: "632873144b0bbfc10ae19430",
                __v: 0,
                name: "Mystery",
            },
            author: {
                _id: "632873e706fe265eaee77de9",
                __v: 0,
                name: "Agatha Christie",
            },
            numberInStock: 2,
            rating: 3.83,
            yearPublished: 1945,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1389733809i/16319.jpg",
            description:
                "It’s seven in the morning. The Bantrys wake to find the body of a young woman in their library. She is wearing evening dress and heavy make-up, which is now smeared across her cheeks.\n\nBut who is she? How did she get there? And what is the connection with another dead girl, whose charred remains are later discovered in an abandoned quarry?\n\nThe respectable Bantrys invite Miss Marple to solve the mystery… before tongues start to wag.",
            __v: 0,
            numberRentedOut: 0,
        },
        {
            _id: "651765abb66bbab77eb9ed5b",
            title: "Murder on the Orient Express",
            genre: {
                _id: "632873144b0bbfc10ae19430",
                __v: 0,
                name: "Mystery",
            },
            author: {
                _id: "632873e706fe265eaee77de9",
                __v: 0,
                name: "Agatha Christie",
            },
            numberInStock: 0,
            rating: 4.2,
            yearPublished: 1934,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1486131451i/853510.jpg",
            description:
                "Just after midnight, a snowdrift stops the famous Orient Express in its tracks as it travels through the mountainous Balkans. The luxurious train is surprisingly full for the time of the year but, by the morning, it is one passenger fewer. An American tycoon lies dead in his compartment, stabbed a dozen times, his door locked from the inside.\n\nOne of the passengers is none other than detective Hercule Poirot. On vacation.\n\nIsolated and with a killer on board, Poirot must identify the murderer—in case he or she decides to strike again.",
            __v: 0,
            numberRentedOut: 2,
        },
        {
            _id: "651760180012d3fa25b3ddba",
            title: "The Code of the Woosters",
            genre: {
                _id: "632873144b0bbfc10ae1942e",
                __v: 0,
                name: "Comedy",
            },
            author: {
                _id: "632873e706fe265eaee77de8",
                __v: 0,
                name: "P.G Wodehouse",
            },
            numberInStock: 2,
            rating: 4.32,
            yearPublished: 1938,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1302053718i/9858081.jpg",
            description:
                "Take Gussie Fink-Nottle, Madeline Bassett, old Pop Bassett, the unscrupulous Stiffy Byng, the Rev., an 18th-century cow-creamer, a small brown leather covered notebook and mix with a dose of the aged aunt Dahlia and one has a dangerous brew which spells toil and trouble for Bertie and Jeeves.",
            __v: 0,
            numberRentedOut: 0,
        },
        {
            _id: "651486abc59681a0aa4771ed",
            title: "The Kite Runner",
            genre: {
                _id: "632873144b0bbfc10ae19431",
                __v: 0,
                name: "Fiction",
            },
            author: {
                _id: "63623b7fe54d96d5cc870c5a",
                __v: 0,
                name: "Khaled Hosseini",
            },
            numberInStock: 5,
            rating: 4.34,
            yearPublished: 2003,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1579036753i/77203.jpg",
            description:
                "1970s Afghanistan: Twelve-year-old Amir is desperate to win the local kite-fighting tournament and his loyal friend Hassan promises to help him. But neither of the boys can foresee what would happen to Hassan that afternoon, an event that is to shatter their lives. After the Russians invade and the family is forced to flee to America, Amir realises that one day he must return to an Afghanistan under Taliban rule to find the one thing that his new world cannot grant him: redemption.",
            __v: 0,
            numberRentedOut: 0,
        },
        {
            _id: "651485f0c59681a0aa4771df",
            title: "Inheritance",
            genre: {
                _id: "6363a31891156291b10c0fea",
                __v: 0,
                name: "Fantasy",
            },
            author: {
                _id: "632873e706fe265eaee77deb",
                __v: 0,
                name: "Christopher Paolini",
            },
            numberInStock: 3,
            rating: 4.13,
            yearPublished: 2011,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1390886502i/7664041.jpg",
            description:
                "It began with Eragon... It ends with Inheritance.\n\nNot so very long ago, Eragon — Shadeslayer, Dragon Rider — was nothing more than a poor farm boy, and his dragon, Saphira, only a blue stone in the forest. Now the fate of an entire civilization rests on their shoulders.\n\nLong months of training and battle have brought victories and hope, but they have also brought heartbreaking loss. And still, the real battle lies ahead: they must confront Galbatorix. When they do, they will have to be strong enough to defeat him. And if they cannot, no one can. There will be no second chance.\n\nThe Rider and his dragon have come further than anyone dared to hope. But can they topple the evil king and restore justice to Alagaësia? And if so, at what cost?\n\nThis is the spellbinding conclusion to Christopher Paolini's worldwide bestselling Inheritance cycle.",
            __v: 0,
            numberRentedOut: 1,
        },
        {
            _id: "651452f1c59681a0aa477199",
            title: "Brisinger",
            genre: {
                _id: "6363a31891156291b10c0fea",
                __v: 0,
                name: "Fantasy",
            },
            author: {
                _id: "632873e706fe265eaee77deb",
                __v: 0,
                name: "Christopher Paolini",
            },
            numberInStock: 4,
            rating: 4.08,
            yearPublished: 2008,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1661030683i/2248573.jpg",
            description:
                "It's been only months since Eragon first uttered \"brisingr\", an ancient language term for fire. Since then, he's not only learned to create magic with words — he's been challenged to his very core. Following the colossal battle against the Empires warriors on the Burning Plains, Eragon and his dragon, Saphira, have narrowly escaped with their lives. Still, there is more adventure at hand for the Rider and his dragon, as Eragon finds himself bound by a tangle of promises he may not be able to keep.\n\nFirst is Eragon's oath to his cousin, Roran: to help rescue Roran's beloved from King Galbatorix's clutches. But Eragon owes his loyalty to others, too. The Varden are in desperate need of his talents and strength — as are the elves and dwarves. When unrest claims the rebels and danger strikes from every corner, Eragon must make choices — choices that will take him across the Empire and beyond, choices that may lead to unimagined sacrifice.\n\nEragon is the greatest hope to rid the land of tyranny. Can this once simple farm boy unite the rebel forces and defeat the king?",
            __v: 0,
            numberRentedOut: 0,
        },
        {
            _id: "65145251c59681a0aa47718b",
            title: "Eldest",
            genre: {
                _id: "6363a31891156291b10c0fea",
                __v: 0,
                name: "Fantasy",
            },
            author: {
                _id: "632873e706fe265eaee77deb",
                __v: 0,
                name: "Christopher Paolini",
            },
            numberInStock: 2,
            rating: 4.02,
            yearPublished: 2005,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1387119654i/45978.jpg",
            __v: 0,
            numberRentedOut: 0,
        },
        {
            _id: "65145107c59681a0aa477174",
            title: "Charlie and the great glass elevator",
            genre: {
                _id: "635a3c7e73691fb5ae1cecd5",
                __v: 0,
                name: "Children",
            },
            author: {
                _id: "635a3d3d73691fb5ae1ced74",
                __v: 0,
                name: "Roald Dahl",
            },
            numberInStock: 5,
            rating: 3.68,
            yearPublished: 1972,
            coverImage:
                "https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1309211607i/6687.jpg",
            description:
                "Last seen flying through the sky in a giant elevator in Charlie and the Chocolate Factory, Charlie Bucket's back for another adventure. When the giant elevator picks up speed, Charlie, Willy Wonka, and the gang are sent hurtling through space and time. Visiting the world’' first space hotel, battling the dreaded Vermicious Knids, and saving the world are only a few stops along this remarkable, intergalactic joyride.",
            __v: 0,
            numberRentedOut: 0,
        },
    ],
    setBooks: jest.fn(),
    isLoading: false,
    error: "",
    setError: jest.fn(),
});

export default useBooks;
