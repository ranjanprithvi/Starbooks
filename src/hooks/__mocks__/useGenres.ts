const useGenres = jest.fn().mockReturnValue({
    genres: [
        { _id: "1", name: "genre1" },
        { _id: "2", name: "genre2" },
    ],
    setGenres: jest.fn(),
    error: "",
    setError: jest.fn(),
    isLoading: false,
});

export default useGenres;
