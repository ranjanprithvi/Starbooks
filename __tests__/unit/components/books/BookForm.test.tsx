import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import apiClient from "../../../../src/services/api-client";
// import { useToast } from "../../../../src/hooks/generic/useToast";
import React from "react";
import { bookObj } from "../../../../src/__mocks__/data";
import { MemoryRouter } from "react-router-dom";
import Routes from "../../../../src/Routes";
import _ from "lodash";

//mock apiClient
jest.mock("../../../../src/services/api-client");
const mApiClient = apiClient as jest.Mocked<typeof apiClient>;
mApiClient.post.mockResolvedValue({ data: bookObj });

jest.mock("../../../../src/hooks/useAuthors");
jest.mock("../../../../src/hooks/useGenres");
jest.mock("../../../../src/hooks/useBook");
const mNavigate = jest.fn();

//mock useToast
const mToast = jest.fn();
jest.mock("@chakra-ui/react", () => ({
    ...(jest.requireActual("@chakra-ui/react") as any),
    useToast: () => mToast,
}));

//mock useNavigate
jest.mock("react-router-dom", () => ({
    ...(jest.requireActual("react-router-dom") as any),
    useNavigate: () => mNavigate,
}));

jest.mock("../../../../src/components/common/ProtectedAdminComponent");

describe("BookForm", () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders", async () => {
        const tree = renderer
            .create(
                <MemoryRouter initialEntries={[`/books/123`]}>
                    <Routes />
                </MemoryRouter>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("renders with none of the fields populated when creating a book", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );
        expect(screen.getByRole("heading")).toHaveTextContent("New Book");
        expect(screen.getByLabelText(/title/i)).toHaveValue("");
        expect(screen.getByLabelText(/author/i)).toHaveValue("");
        expect(screen.getByLabelText(/genre/i)).toHaveValue("");
        expect(screen.getByLabelText(/year published/i)).toHaveValue(null);
        expect(screen.getByLabelText(/rating/i)).toHaveValue(null);
        expect(screen.getByLabelText(/number in stock/i)).toHaveValue(null);
        expect(screen.getByLabelText(/cover image/i)).toHaveValue("");
        expect(screen.getByLabelText(/description/i)).toHaveValue("");
    });

    it("renders with all the fields populated when editing a book", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );

        expect(screen.getByRole("heading")).toHaveTextContent("Edit Book");
        expect(screen.getByLabelText(/title/i)).toHaveValue(bookObj.title);
        expect(screen.getByLabelText(/author/i)).toHaveValue(
            bookObj.author._id
        );
        expect(screen.getByLabelText(/genre/i)).toHaveValue(bookObj.genre._id);
        expect(screen.getByLabelText(/year published/i)).toHaveValue(
            bookObj.yearPublished
        );
        expect(screen.getByLabelText(/rating/i)).toHaveValue(bookObj.rating);
        expect(screen.getByLabelText(/number in stock/i)).toHaveValue(
            bookObj.numberInStock
        );
        expect(screen.getByLabelText(/cover image/i)).toHaveValue(
            bookObj.coverImage
        );
        expect(screen.getByLabelText(/description/i)).toHaveValue(
            bookObj.description
        );
    });

    it("calls the post method with the right data when all fields are populated", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );

        await act(async () => {
            fireEvent.change(screen.getByLabelText(/title/i), {
                target: { value: bookObj.title },
            });
            fireEvent.change(
                screen.getByRole("combobox", { name: /author/i }),
                { target: { value: bookObj.author._id } }
            );
            fireEvent.change(screen.getByRole("combobox", { name: /genre/i }), {
                target: { value: bookObj.genre._id },
            });
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: bookObj.yearPublished },
            });
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: bookObj.rating },
            });
            fireEvent.change(screen.getByLabelText(/number in stock/i), {
                target: { value: bookObj.numberInStock },
            });
            fireEvent.change(screen.getByLabelText(/cover image/i), {
                target: { value: bookObj.coverImage },
            });
            fireEvent.change(screen.getByLabelText(/description/i), {
                target: { value: bookObj.description },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        const bookData = {
            ..._.omit(bookObj, ["_id"]),
            author: bookObj.author._id,
            genre: bookObj.genre._id,
        };
        expect(mApiClient.post).toHaveBeenCalledWith("/books", bookData);
        expect(mNavigate).toHaveBeenCalledWith("/bookDetails/123", {
            replace: true,
        });
    });

    it("does not submit when no title, author, genre, year published is entered", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(screen.getByText(/title is required/i)).toBeInTheDocument();
        expect(screen.getByText(/author is required/i)).toBeInTheDocument();
        expect(screen.getByText(/genre is required/i)).toBeInTheDocument();
        expect(
            screen.getByText(/year published is required/i)
        ).toBeInTheDocument();
    });

    it("does not submit when title is less than 3 characters long", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/title/i), {
                target: { value: "ab" },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.patch).not.toHaveBeenCalled();
        expect(
            screen.getByText(/title must be at least 3 characters long/i)
        ).toBeInTheDocument();
    });

    it("does not submit when year published is a negative number", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: -1 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.patch).not.toHaveBeenCalled();
        expect(
            screen.getByText(/Year published must be a positive number/i)
        ).toBeInTheDocument();
    });

    it("does not submit when year published is not a whole number", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: 2000.5 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/Year published must be a whole number/i)
        ).toBeInTheDocument();
    });

    it("does not submit when year published is in the future", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: 3000 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(
                /Year published must be less than or equal to current year/i
            )
        ).toBeInTheDocument();
    });

    it("does not submit when rating is less than 0", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: -1 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/Rating must be between 0 and 5/i)
        ).toBeInTheDocument();
    });

    it("does not submit when rating is greater than 5", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: 6 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/Rating must be between 0 and 5/i)
        ).toBeInTheDocument();
    });

    it("does not submit when rating is not a multiple of 0.01", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: 3.001 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/Rating must be a multiple of 0.01/i)
        ).toBeInTheDocument();
    });

    it("does not submit when number in stock is not a whole number", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/number in stock/i), {
                target: { value: 2.5 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/number in stock must be a whole number/i)
        ).toBeInTheDocument();
    });

    it("does not submit when number in stock is negative", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/number in stock/i), {
                target: { value: -2 },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.post).not.toHaveBeenCalled();
        expect(
            screen.getByText(/number in stock must be non-negative/i)
        ).toBeInTheDocument();
    });

    it("does not submit when coverimage is not a valid url", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/123`]}>
                <Routes />
            </MemoryRouter>
        );
        await act(async () => {
            fireEvent.change(screen.getByLabelText(/cover image/i), {
                target: { value: "abasdf" },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });

        expect(mApiClient.patch).not.toHaveBeenCalled();
        expect(
            screen.getByText(/cover image must be a valid url/i)
        ).toBeInTheDocument();
    });

    it.skip("displays error message in case of bad request", async () => {
        const message = "Invalid username or password";
        mApiClient.post.mockRejectedValueOnce({
            response: {
                data: message,
            },
        });

        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );

        await act(async () => {
            fireEvent.change(screen.getByLabelText(/title/i), {
                target: { value: bookObj.title },
            });
            fireEvent.change(
                screen.getByRole("combobox", { name: /author/i }),
                { target: { value: bookObj.author._id } }
            );
            fireEvent.change(screen.getByRole("combobox", { name: /genre/i }), {
                target: { value: bookObj.genre._id },
            });
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: bookObj.yearPublished },
            });
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: bookObj.rating },
            });
            fireEvent.change(screen.getByLabelText(/number in stock/i), {
                target: { value: bookObj.numberInStock },
            });
            fireEvent.change(screen.getByLabelText(/cover image/i), {
                target: { value: bookObj.coverImage },
            });
            fireEvent.change(screen.getByLabelText(/description/i), {
                target: { value: bookObj.description },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });
        // expect(
        //     await screen.findByText(/invalid username or password/i)
        // ).toBeInTheDocument();
    });

    it("displays error message in case of network error", async () => {
        const message = "Sorry. Something went wrong";
        mApiClient.post.mockResolvedValueOnce({
            data: { token: "1234", isAdmin: true },
        });
        process.env.VITE_BACKEND_URL = "http://invalidUrl";
        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );

        await act(async () => {
            fireEvent.change(screen.getByLabelText(/title/i), {
                target: { value: bookObj.title },
            });
            fireEvent.change(
                screen.getByRole("combobox", { name: /author/i }),
                { target: { value: bookObj.author._id } }
            );
            fireEvent.change(screen.getByRole("combobox", { name: /genre/i }), {
                target: { value: bookObj.genre._id },
            });
            fireEvent.change(screen.getByLabelText(/year published/i), {
                target: { value: bookObj.yearPublished },
            });
            fireEvent.change(screen.getByLabelText(/rating/i), {
                target: { value: bookObj.rating },
            });
            fireEvent.change(screen.getByLabelText(/number in stock/i), {
                target: { value: bookObj.numberInStock },
            });
            fireEvent.change(screen.getByLabelText(/cover image/i), {
                target: { value: bookObj.coverImage },
            });
            fireEvent.change(screen.getByLabelText(/description/i), {
                target: { value: bookObj.description },
            });

            const submitButton = screen.getByText(
                /submit/i
            ) as HTMLButtonElement;
            submitButton.click();
        });
    });

    it("navigates to previous screen when canceled", async () => {
        render(
            <MemoryRouter initialEntries={[`/books/new`]}>
                <Routes />
            </MemoryRouter>
        );

        const cancelButton = screen.getByText(/cancel/i);
        cancelButton.click();
        expect(mNavigate).toHaveBeenCalledWith(-1);
    });
});
