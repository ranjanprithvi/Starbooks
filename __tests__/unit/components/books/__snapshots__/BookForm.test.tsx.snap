// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookForm renders 1`] = `
<div
  className="css-3xydd0"
>
  <div
    className="css-zmtksu"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="chakra-stack css-tl3ftk"
      >
        <h2
          className="chakra-heading css-r7q7qr"
        >
          Edit Book
        </h2>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="title"
              id="field-:r1:-label"
            >
              Title
            </label>
            <input
              className="chakra-input css-0"
              disabled={false}
              id="title"
              name="title"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onWheel={[Function]}
              readOnly={false}
              required={false}
              step="any"
            />
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="author"
              id="field-:r3:-label"
            >
              Author
            </label>
            <div
              className="chakra-select__wrapper css-42b2qy"
            >
              <select
                className="chakra-select css-1ik61og"
                disabled={false}
                id="author"
                name="author"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={false}
                required={false}
              >
                <option
                  value=""
                >
                  --Select author--
                </option>
                <option
                  value="1"
                >
                  author1
                </option>
                <option
                  value="2"
                >
                  author2
                </option>
              </select>
              <div
                className="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-select__icon"
                  focusable={false}
                  role="presentation"
                  style={
                    {
                      "color": "currentColor",
                      "height": "1em",
                      "width": "1em",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="genre"
              id="field-:r5:-label"
            >
              Genre
            </label>
            <div
              className="chakra-select__wrapper css-42b2qy"
            >
              <select
                className="chakra-select css-1ik61og"
                disabled={false}
                id="genre"
                name="genre"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={false}
                required={false}
              >
                <option
                  value=""
                >
                  --Select genre--
                </option>
                <option
                  value="1"
                >
                  genre1
                </option>
                <option
                  value="2"
                >
                  genre2
                </option>
                <option
                  value="3"
                >
                  genre3
                </option>
                <option
                  value="4"
                >
                  genre4
                </option>
              </select>
              <div
                className="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden={true}
                  className="chakra-select__icon"
                  focusable={false}
                  role="presentation"
                  style={
                    {
                      "color": "currentColor",
                      "height": "1em",
                      "width": "1em",
                    }
                  }
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="yearPublished"
              id="field-:r7:-label"
            >
              Year Published
            </label>
            <input
              className="chakra-input css-0"
              disabled={false}
              id="yearPublished"
              name="yearPublished"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onWheel={[Function]}
              readOnly={false}
              required={false}
              step="any"
              type="number"
            />
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="rating"
              id="field-:r9:-label"
            >
              Rating
            </label>
            <input
              className="chakra-input css-0"
              disabled={false}
              id="rating"
              name="rating"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onWheel={[Function]}
              readOnly={false}
              required={false}
              step="any"
              type="number"
            />
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="numberInStock"
              id="field-:rb:-label"
            >
              Number in Stock
            </label>
            <input
              className="chakra-input css-0"
              disabled={false}
              id="numberInStock"
              name="numberInStock"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onWheel={[Function]}
              readOnly={false}
              required={false}
              step="any"
              type="number"
            />
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="coverImage"
              id="field-:rd:-label"
            >
              Cover Image
            </label>
            <input
              className="chakra-input css-0"
              disabled={false}
              id="coverImage"
              name="coverImage"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onWheel={[Function]}
              readOnly={false}
              required={false}
              step="any"
            />
          </div>
        </div>
        <div
          className="chakra-form-control css-h82jag"
          role="group"
        >
          <div
            className="chakra-form-control css-0"
            role="group"
          >
            <label
              className="chakra-form__label css-uvho5g"
              htmlFor="description"
              id="field-:rf:-label"
            >
              Description
            </label>
            <textarea
              className="chakra-textarea css-0"
              disabled={false}
              id="description"
              name="description"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={false}
              required={false}
            />
          </div>
        </div>
        <div
          className="chakra-stack css-mogqpi"
        >
          <button
            className="chakra-button css-4xx2wk"
            id="submit"
            type="submit"
          >
            Submit
          </button>
          <button
            className="chakra-button css-4xx2wk"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
